client/client.go

```diff
@@ -1115,91 +1113,121 @@ func (r *NotaryRepository) bootstrapClient(checkInitialized bool) (*TUFClient, e
+func (r *NotaryRepository) pubKeyListForRotation(role data.RoleName, serverManaged bool, privKeys []string, pubKeys data.KeyList) (pubKeyList data.KeyList, err error) {
+	var pubKey data.PublicKey
+	switch {
+	case serverManaged:
+		// If server manages the key being rotated, request a rotation and return the new key
 		remote := r.getRemoteStore()
 		pubKey, err = rotateRemoteKey(role, remote)
-		pubKeyList = make(data.KeyList, 0, 1)
 		pubKeyList = append(pubKeyList, pubKey)
-		if err != nil {
-			return nil, fmt.Errorf("unable to rotate remote key: %s", err)
+		return pubKeyList, err
+	case len(pubKeys) == 0 && len(privKeys) == 0:
+		if pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey); err != nil {
+			return nil, fmt.Errorf("unable to generate key: %s", err)
 		}
-		return pubKeyList, nil
-	}
-
-	// If no new keys are passed in, we generate one
-	if len(newKeys) == 0 {
-		pubKeyList = make(data.KeyList, 0, 1)
-		pubKey, err = r.CryptoService.Create(role, r.gun, data.ECDSAKey)
 		pubKeyList = append(pubKeyList, pubKey)
-	}
-	if err != nil {
-		return nil, fmt.Errorf("unable to generate key: %s", err)
+	case len(privKeys) != 0:
+		if pubKeyList, err = r.pubKeyFromPrivKIDs(privKeys, pubKeys); err != nil {
+			return nil, err
+		}
+	default:
+		pubKeyList = pubKeys
 	}
 
-	// If a list of keys to rotate to are provided, we add those
-	if len(newKeys) > 0 {
-		pubKeyList = make(data.KeyList, 0, len(newKeys))
-		for _, keyID := range newKeys {
-			pubKey = r.CryptoService.GetKey(keyID)
-			if pubKey == nil {
-				return nil, fmt.Errorf("unable to find key: %s", keyID)
+	return r.pubKeysToCerts(role, pubKeyList)
+}
+
+// privateKeyIDsMatchPublicKeys checks each KeyID for corresponding public key and returns them in a list
+func (r *NotaryRepository) pubKeyFromPrivKIDs(privKIDs []string, pubKeys data.KeyList) (data.KeyList, error) {
+	if len(pubKeys) == 0 {
+		var ids []data.PublicKey
+		for _, id := range privKIDs {
+			var pubKey data.PublicKey
+			if pubKey = r.CryptoService.GetKey(id); pubKey == nil {
+				return nil, fmt.Errorf("unable to find key: %s", id)
 			}
-			pubKeyList = append(pubKeyList, pubKey)
+			ids = append(ids, pubKey)
 		}
+		return ids, nil
 	}
 
-	// Convert to certs (for root keys)
-	if pubKeyList, err = r.pubKeysToCerts(role, pubKeyList); err != nil {
-		return nil, err
+	// check if each private key has a matching public key in 'pubKeys' list
+	// if all private keys are matched and there are unmatched public keys
+	// then those public keys are assumed to be valid as well.
+	pubKeyIDs := make(map[string]bool)
+	for _, k := range pubKeys {
+		kid, err := utils.CanonicalKeyID(k)
+		if err != nil {
+			return nil, fmt.Errorf("invalid public key: %v", err)
+		}
+		pubKeyIDs[kid] = true
 	}
-
-	return pubKeyList, nil
+	for _, id := range privKIDs {
+		if exist := pubKeyIDs[id]; !exist {
+			return nil, fmt.Errorf("private key %v require matching public key", id)
```

Comment:
Thanks for the countless reviews and tips @riyazdf 😄 . I really appreciate it! 

I have corrected the typos and squashed everything into one commit.